include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+	// TODO: insert utility functions:  validate(), print, ostream<<, errors(), warnings()...
+	//
+
+private:
+	const rclcpp::Node::SharedPtr node_;
+	const std::string base_name_;
+
+	// Class similar to the one in rosparam_shortcuts/node_parameters.h
+	// implements ROS calls/checks/callbacks for node_parameters_
+	NodeParametersInterface node_parameters_interface_;
+
+	// Store parameters, only node parameters (rclcpp::Parameters) are stored inside NodeConfig
+	std::map<std::string, Parameter&> parameters_;
+	std::map<std::string, std::shared_ptr<NodeParameter>> node_parameters_;
+
+	// Needed? keep track of registered callbacks that are required for linking SetParameters requests or
+	// exposing high-level update changes (invalid/valid states)
+	// std::string<std::string, std::string> registered_node_parameter_callbacks_ids_;
+	// std::string<std::string, std::string> registered_parameter_callbacks_ids_;
+}
+
+struct ExampleConfig : public NodeConfig
+{
+	ExampleConfig(const rclcpp::Node::SharedPtr& node) : NodeConfig(node) {
+		// declare simple parameters just like that
+		declare(int1_param);
+		declare(int2_param);
+		declare(int3_param);
+
+		// "double_param", add some description and validity checks
+		declare(double_param)
+		    .describe("This is a very important double parameter", "should be greater than 0 and not 7")
+		    .warnIf([](const double& val, std::string& message) {
+			    message = "Value is 7, are you sure this is a good value?";
+			    return val == 7;
+		    })
+		    .errorIf([](const double& val, std::string& message) {
+			    message = "Value must not be negative!";
+			    return val < 0;
+		    });
+
+		// "string_param", use a parameter error condition "EMPTY"
+		declare(string_param)
+		    .describe("meh")
+		    .warnIf([](const std::string& val, std::string& message) {
+			    message = "string is much too long";
+			    return val.size() > 100;
+		    })
+		    .errorIf(EMPTY);
+
+		// "double_map_param", validate keys and values, reject if not valid
+		declare(double_map_param).describe("6-dof joint state").errorIf([](const auto& values, std::string& message) {
+			message = "Invalid joint state map values, expected 6-dof";
+			return values.size() != 6 || checkJointNamesInRobotModel(values);  // from somewhere
+		});
+
+		//
+		// can already check for errors/warnings here...
+		//
+	}
+
+	Parameter<double> double_param{ "double_param" };
+	Parameter<std::vector<std::string>> string_param{ "string_param" };
+	Parameter<std::map<std::string, double>> double_map_param{ "double_map_param" };
+	Parameter<int> int1_param{ "int1_param" };
+	Parameter<int> int2_param{ "int2_param" };
+	Parameter<int> int3_param{ "int3_param" };
+}
+
+void useConfig(const rclcpp::Node::SharedPtr& node)
+{
+	// use non-const config for initializing callbacks
+	// For safety, it should only be accessed as const in nodes
+	const ExampleConfig config(node);
+
+	// access copy of value
+	double value = config.double_param();
```

Comment:
This is a nice interface for getting the data.  I'm assuming this will lock a mutex to make sure when you are reading it isn't also being updated.  Maybe this should throw an exception if the value has not been initialized (no default and value wasn't set).  Maybe there should be a `read(timeout)` method for the parameters with some default value set for the timeout so you can specify the timeout if you want to per-parameter.

That might not be a good idea, we need to keep the interface simple and make it easy to follow a good pattern with it.

Reply:
Yeah, the read is locked by a shared mutex. I don't know if a timeout is really needed at all. The only scenario I see is if that some thread locks write access and doesn't release the mutex for a long time, which would be a user error that definitely shouldn't happen. I guess that if we really wanted to protect for that, we could use an internal timeout that throws an exception if the value has been locked for too long.